src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/TemporaryArray`1.cs

```diff
@@ -0,0 +1,308 @@
+            {
+                _builder.AddRange(items);
+            }
+            else if (_count + items.Length <= InlineCapacity)
+            {
+                foreach (var item in items)
+                {
+                    // Increase _count before assigning values since the indexer has a bounds check.
+                    _count++;
+                    this[_count - 1] = item;
+                }
+            }
+            else
+            {
+                MoveInlineToBuilder();
+                _builder.AddRange(items);
+            }
+        }
+
+        public void Clear()
+        {
+            if (_builder is not null)
+            {
+                // Keep using a builder even if we now fit in inline storage to avoid churn in the object pools.
+                _builder.Clear();
+            }
+            else
+            {
+                this = Empty;
+            }
+        }
+
+        public readonly Enumerator GetEnumerator()
+        {
+            return new Enumerator(in this);
+        }
+
+        /// <summary>
+        /// Create an <see cref="ImmutableArray{T}"/> with the elements currently held in the temporary array, and clear
+        /// the array.
+        /// </summary>
+        /// <returns></returns>
+        public ImmutableArray<T> MoveToImmutable()
+        {
+            if (_builder is not null)
+            {
+                return _builder.MoveToImmutable();
+            }
+            else
+            {
+                var result = _count switch
+                {
+                    0 => ImmutableArray<T>.Empty,
+                    1 => ImmutableArray.Create(_item0),
+                    2 => ImmutableArray.Create(_item0, _item1),
+                    3 => ImmutableArray.Create(_item0, _item1, _item2),
+                    4 => ImmutableArray.Create(_item0, _item1, _item2, _item3),
+                    _ => throw ExceptionUtilities.Unreachable,
+                };
+
+                // Since _builder is null on this path, we can overwrite the whole structure to Empty to reset all
+                // inline elements to their default value and the _count to 0.
+                this = Empty;
+
+                return result;
+            }
+        }
+
+        /// <summary>
+        /// Transitions the current <see cref="TemporaryArray{T}"/> from inline storage to dynamic storage storage. An
+        /// <see cref="ArrayBuilder{T}"/> instance is taken from the shared pool, and all elements currently in inline
+        /// storage are added to it. After this point, dynamic storage will be used instead of inline storage.
+        /// </summary>
+        [MemberNotNull(nameof(_builder))]
+        private void MoveInlineToBuilder()
+        {
+            Debug.Assert(_builder is null);
+
+            var builder = ArrayBuilder<T>.GetInstance();
+            for (var i = 0; i < _count; i++)
+            {
+                builder.Add(this[i]);
+                this[i] = default!;
```

Comment:
👍  up to you if we should explicitly call this out

Reply:
I wrapped it in `RuntimeHelpers.IsReferenceOrContainsReferences<T>()` so it's clearer why this is getting set.